Setup/UpgradeData.php

```diff
@@ -141,4 +156,54 @@ private function insertStoreDomain()
             }
         }
     }
+
+    /**
+     * Converts serialized product data to base64 encoded data to avoid charset & collation problems
+     *
+     * @param AdapterInterface $connection
+     */
+    public function convertProductDataToBase64(AdapterInterface $connection)
+    {
+        try {
+            $cachedProductCollection = $this->cacheCollectionFactory->create()->setPageSize(1000);
+            $iterator = new PagingIterator($cachedProductCollection);
+            foreach ($iterator as $page) {
+                $canBeConverted = [];
+                $nullableIds = [];
+                /* @var Cache $cachedProduct */
+                foreach ($page as $cachedProduct) {
+                    try {
+                        unserialize($cachedProduct->getProductData(), [Product::class]); // @codingStandardsIgnoreLine
+                        $canBeConverted[] = $cachedProduct->getId();
+                    } catch (\Exception $e) {
+                        $nullableIds[] = $cachedProduct->getId();
+                    }
+                }
+                if (!empty($cachedProduct)) {
+                    $convertSql = sprintf(
+                        'UPDATE %s SET %s = TO_BASE64(%s) WHERE %s IN(%s)',
+                        CacheResource::TABLE_NAME,
+                        Cache::PRODUCT_DATA,
+                        Cache::PRODUCT_DATA,
+                        Cache::ID,
+                        implode(',', $canBeConverted)
+                    );
+                    $connection->query($convertSql); // @codingStandardsIgnoreLine
+                }
+                if (!empty($nullableIds)) {
+                    $setNullSql = sprintf(
+                        'UPDATE %s SET %s = NULL, %s=1 WHERE %s IN(%s)',
+                        CacheResource::TABLE_NAME,
+                        Cache::PRODUCT_DATA,
+                        Cache::IS_DIRTY,
+                        Cache::ID,
+                        implode(',', $nullableIds)
+                    );
+                    $connection->query($setNullSql);  // @codingStandardsIgnoreLine
+                }
```

Comment:
I'd wrap those 2 if cases into small private functions to reduce this functions size